### PR TITLE
HIVE-22267 : Support password based authentication for HMS along-side kerberos authentication.

### PR DESCRIPTION
**Pull request by ashutosh-bapat**

### CODE REVIEW

File: itests/hive-unit-hadoop2/src/test/java/org/apache/hadoop/hive/metastore/security/TestHadoopAuthBridge23.java [hunk 1]
```diff
@@ -94,16 +94,15 @@ public Server() throws TTransportException {
...
-      throws TTransportException {
+      public TSaslServerTransport.Factory createSaslServerTransportFactory(Map<String, String> saslProps) {
         TSaslServerTransport.Factory transFactory =
           new TSaslServerTransport.Factory();
         transFactory.addServerDefinition(AuthMethod.DIGEST.getMechanismName(),
             null, SaslRpcServer.SASL_DEFAULT_REALM,
             saslProps,
             new SaslDigestCallbackHandler(secretManager));
 
-        return new TUGIAssumingTransportFactory(transFactory, realUgi);
```
Comment (maheshk114) [inline itests/hive-unit-hadoop2/src/test/java/org/apache/hadoop/hive/metastore/security/TestHadoopAuthBridge23.java]: why this change is required ?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreAuthenticationProviderFactory.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    throws AuthenticationException {
+    return getAuthenticationProvider(new Configuration(), authMethod);
+  }
+
+  public static MetaStorePasswdAuthenticationProvider getAuthenticationProvider(Configuration conf, AuthMethods authMethod)
+    throws AuthenticationException {
+    if (authMethod == AuthMethods.LDAP) {
+      return new MetaStoreLdapAuthenticationProviderImpl(conf);
+    } else if (authMethod == AuthMethods.CUSTOM) {
+      return new MetaStoreCustomAuthenticationProviderImpl(conf);
```
Comment (maheshk114) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreAuthenticationProviderFactory.java]: PAM is not supported ?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  @SuppressWarnings("unchecked")
+  MetaStoreConfigAuthenticationProviderImpl(Configuration conf) throws AuthenticationException {
+    userName = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.THRIFT_AUTH_CONFIG_USERNAME);
+    password = MetastoreConf.getVar(conf, MetastoreConf.ConfVars.THRIFT_AUTH_CONFIG_PASSWORD);
+
+    if (null == userName || userName.isEmpty()) {
+      throw new AuthenticationException("No username specified in " + MetastoreConf.ConfVars.THRIFT_AUTH_CONFIG_USERNAME);
+    }
+
+    if (null == password || password.isEmpty()) {
```
Comment (maheshk114) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java]: password can be empty

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      throw new AuthenticationException("No password specified in " + MetastoreConf.ConfVars.THRIFT_AUTH_CONFIG_USERNAME);
+    }
+  }
+
+  @Override
+  public void Authenticate(String authUser, String authPassword) throws AuthenticationException {
+    if (!userName.equals(authUser)) {
+      throw new AuthenticationException("Invalid user " + authUser);
```
Comment (maheshk114) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java]: dont put username and password in exception ..add it as a debug log

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 5]
```diff
@@ -629,6 +635,97 @@ public static ConfVars getMetaConf(String name) {
         "hive-metastore/_HOST@EXAMPLE.COM",
         "The service principal for the metastore Thrift server. \n" +
             "The special string _HOST will be replaced automatically with the correct host name."),
+    THRIFT_METASTORE_AUTHENTICATION("metastore.authentication", "hive.metastore.authentication",
+            "NOSASL",
+      new StringSetValidator("NOSASL", "NONE", "LDAP", "KERBEROS", "CUSTOM"),
+        "Client authentication types.\n" +
+                "  NONE: no authentication check\n" +
+                "  LDAP: LDAP/AD based authentication\n" +
+                "  KERBEROS: Kerberos/GSSAPI authentication\n" +
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java]: none means that the SASL protocol will be executed but there will be no authentication check. The user will be simply accepted as is.

NoSASL means there's no SASL protocol, no user, just plain transport.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 6]
```diff
@@ -1039,8 +1136,27 @@ public static ConfVars getMetaConf(String name) {
...
+            "metastore.client.use.plain.auth", false,
+            "If true, clients will authenticate using plain authentication, by providing username" +
+                    " and password."),
+    METASTORE_CLIENT_PLAIN_USERNAME("metastore.client.plain.username",
+            "metastore.client.plain.username",  "",
+        "The username used by the metastore client when " +
+                    METASTORE_CLIENT_USE_PLAIN_AUTH + " is true. The password is obtained from " +
+                    CredentialProviderFactory.CREDENTIAL_PROVIDER_PATH + " using username as the " +
+                    "alias."),
+    THRIFT_AUTH_CONFIG_USERNAME("metastore.authentication.config.username",
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java]: The first one is client side config. It will be used whenever there's PLAIN authentication. That specifies what username to provide to the server.

The other is server side and used only when CONFIG authentication is used. It's the username that will be matched against the username provided by the client.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 7]
```diff
@@ -590,7 +592,30 @@ private void open() throws MetaException {
             transport = new TSocket(store.getHost(), store.getPort(), clientSocketTimeout);
           }
 
-          if (useSasl) {
+          if (usePasswordAuth) {
+            // we are using PLAIN Sasl connection with user/password
+            LOG.debug("HMSC::open(): Creating plain authentication thrift connection.");
+            String userName = MetastoreConf.getVar(conf, ConfVars.METASTORE_CLIENT_PLAIN_USERNAME);
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: Done.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 8]
```diff
@@ -590,7 +592,30 @@ private void open() throws MetaException {
...
+            // The password is not directly provided. It should be obtained from a keystore pointed
+            // by configuration "hadoop.security.credential.provider.path".
+            try {
+              String passwd = null;
+              char[] pwdCharArray = conf.getPassword(userName);
+              if (null != pwdCharArray) {
+               passwd = new String(pwdCharArray);
+              }
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java]: Done.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 9]
```diff
@@ -629,6 +635,97 @@ public static ConfVars getMetaConf(String name) {
...
+      new StringSetValidator("NOSASL", "NONE", "LDAP", "KERBEROS", "CUSTOM"),
+        "Client authentication types.\n" +
+                "  NONE: no authentication check\n" +
+                "  LDAP: LDAP/AD based authentication\n" +
+                "  KERBEROS: Kerberos/GSSAPI authentication\n" +
+                "  CUSTOM: Custom authentication provider\n" +
+                "          (Use with property metastore.custom.authentication.class)\n" +
+                "  CONFIG: username and password is specified in the config" +
+                "  NOSASL:  Raw transport"),
+    THRIFT_CUSTOM_AUTHENTICATION_CLASS("metastore.custom.authentication.class",
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java]: Done.

File: itests/hive-unit-hadoop2/src/test/java/org/apache/hadoop/hive/metastore/security/TestHadoopAuthBridge23.java (hunk 1 above)
Comment (ashutosh-bapat) [reply itests/hive-unit-hadoop2/src/test/java/org/apache/hadoop/hive/metastore/security/TestHadoopAuthBridge23.java]: Earlier HiveMetaStore was using createTransportFactory to create a transport factory. But now we are not using it. Instead, we are using createSaslServerTransportFactory, which should not create a wrapped transport.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreAuthenticationProviderFactory.java (hunk 2 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreAuthenticationProviderFactory.java]: No. No PAM support.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java (hunk 4 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java]: Done. Added debug logs and invalid credentials exception.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java (hunk 3 above)
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreConfigAuthenticationProviderImpl.java]: Fixed. Allows empty password.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 10]
```diff
@@ -1039,8 +1136,27 @@ public static ConfVars getMetaConf(String name) {
             "More users can be added in ADMIN role later."),
     USE_SSL("metastore.use.SSL", "hive.metastore.use.SSL", false,
         "Set this to true for using SSL encryption in HMS server."),
+    // We should somehow unify next two options.
     USE_THRIFT_SASL("metastore.sasl.enabled", "hive.metastore.sasl.enabled", false,
         "If true, the metastore Thrift interface will be secured with SASL. Clients must authenticate with Kerberos."),
+    METASTORE_CLIENT_USE_PLAIN_AUTH("metastore.client.use.plain.auth",
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java]: Done. But for now, only "PLAIN" will be considered. Any other values is ignored. For deprecation, we will revisit it when we will rethink about the configs, in a short while.

File: itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/RemoteHiveMetaStoreDualAuthTest.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+package org.apache.hive.minikdc;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.metastore.TestRemoteHiveMetaStore;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf.ConfVars;
+import org.junit.Before;
+
+public class RemoteHiveMetaStoreDualAuthTest extends TestRemoteHiveMetaStore {
+  // These names are tied with the .jceks file used by the subclasses. So, do not change those.
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply itests/hive-minikdc/src/test/java/org/apache/hive/minikdc/RemoteHiveMetaStoreDualAuthTest.java]: Done.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreCustomAuth.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    MetastoreConf.setVar(conf, ConfVars.THRIFT_URIS, "thrift://localhost:" + port);
+    MetastoreConf.setBoolVar(conf, ConfVars.METASTORE_CLIENT_USE_PLAIN_AUTH, true);
+    String tmpDir = System.getProperty("build.dir");
+    String credentialsPath = "jceks://file" + tmpDir + "/test-classes/creds/hms_plain_auth_test.jceks";
+    conf.set(CredentialProviderFactory.CREDENTIAL_PROVIDER_PATH, credentialsPath);
+
+    try {
+      MetastoreConf.setVar(conf, ConfVars.METASTORE_CLIENT_PLAIN_USERNAME, wrongUser);
+      HiveMetaStoreClient tmpClient = new HiveMetaStoreClient(conf);
+    } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreCustomAuth.java]: Done. It's always MetaException which wraps the actual exception as "Could not connect to meta store using any of the URIs provided." along with the stack trace. So, instead I have checked parts of string.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreCustomAuth.java [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+      gotException = true;
+    }
+    // Trying to log in using wrong username and password should fail
+    Assert.assertTrue(gotException);
```
[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreCustomAuth.java]: Done.

[the remark this answers is not in the document]
Comment (ashutosh-bapat) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestRemoteHiveMetaStoreCustomAuth.java]: While doing this, I found an existing bug. In HiveMetaStoreClient, it could cause NullPointerException if no TTransportException was raised (e.g. no password provided or userName null or truststore not provided etc.). I have fixed that bug as well.
